segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/handler/PravegaConnectionListener.java

```diff
@@ -144,26 +155,48 @@ public void startListening() {
             workerGroup = new NioEventLoopGroup();
         }
 
+        Channels channels = new Channels();
+
         ServerBootstrap b = new ServerBootstrap();
         b.group(bossGroup, workerGroup)
          .channel(nio ? NioServerSocketChannel.class : EpollServerSocketChannel.class)
          .option(ChannelOption.SO_BACKLOG, 100)
          .handler(new LoggingHandler(LogLevel.INFO))
          .childHandler(new ChannelInitializer<SocketChannel>() {
              @Override
-             public void initChannel(SocketChannel ch) throws Exception {
+             public void initChannel(SocketChannel ch) {
                  ChannelPipeline p = ch.pipeline();
-                 if (sslCtx != null) {
-                     SslHandler handler = sslCtx.newHandler(ch.alloc());
-                     p.addLast(handler);
+
+                 // Add SslHandler to the channel's pipeline, if TLS is enabled.
+                 if (enableTls) {
+                     final SslContext sslContext;
+
+                     if (enableTlsReload) {
+                         channels.add(ch);
+
+                         // Creating an SSL Context for each init ensures that the latest cert and key files are used for
+                         // any new channels.
+                         sslContext = TLSHelper.newServerSslContext(pathToTlsCertFile, pathToTlsKeyFile);
+                         log.debug("Created a new SSL context.");
```

Comment:
It is too expensive to create a new `SslContext` on each request.   The best you can do is, use the watch events to update the cached context, and use the cached context whenever creating a new request.

Reply:
I ran a few tests creating an `io.netty.handler.ssl.SslContext.SslContext` using the `io.netty.handler.ssl.SslContextBuilder.forServer(...)` method used in Pravega, and found that it wasn't expensive - at least in terms of the time it takes to create each instance.  One such test was creating `SslContext` 200 times in a loop. It took ~1.3 second the first time, and 96 ms. on average for the rest of the 199 times. 

The call to create a new SslContext happens once per channel. Also, we don't create channels on a channel-per-request. It is shared across multiple requests. They get closed either at the client application's behest or if an exception/error causes the server to close the channel.  

I suspect that updating the cached context can be risky, especially for in-flight requests. And, there might be different consequences depending on the stage the request is at. I find it easier to articulate and reason about the consequences of TLS certificate change with the simplistic model of handling the certificate modification that is there in the PR code today:

* If TLS reload is enabled:
   * If the certificate changes, existing channels will be closed by the server, resulting in the client library code encountering connection drops. Upon connection drops, the library code automatically retries connecting to the server, and this time a new SslContext built using the latest certificate is used. 
  * Regardless of whether the certificate remains the same or modifies, all new channels are created using a SslContext built from the latest certificate. 

   As a result, a new certificate is immediately applied to all existing and new connections. The need to refresh existing channels stems from the need to handle expired certificates too.

* If TLS reload is disabled:
   * SslContext is created once and the same cached instance is used for every channel.  

What do you say?

Reply:
This is not a good take.  You cannot create an `SslContext` per channel.  Channels are hardly as reusable as you suggest.   A simple `ReentrantReadWriteLock` (fair mode) will suffice to protect the cached SSL context; obtain the (shared) read lock whenever a channel is created, and obtain the write lock on the watch thread (after creating the replacement context but before updating the cache).

Do not close existing channels.

Reply:
Upon a deeper look into the pros and cons of the two approaches, I think the approach you have suggested is simpler. I have modified the code and pushed the changes.